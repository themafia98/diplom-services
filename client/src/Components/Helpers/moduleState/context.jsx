import { createContext } from 'react';

const ModuleContextState = createContext();
const ModuleContextActions = createContext();

export { ModuleContextState, ModuleContextActions };
