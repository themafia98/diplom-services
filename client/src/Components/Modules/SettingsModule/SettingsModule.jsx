import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { connect } from 'react-redux';
import { saveComponentStateAction } from '../../../Redux/actions/routerActions';
import { updateUdata, setStatus } from '../../../Redux/actions/publicActions';
import { Collapse, Switch, Input, Button, message } from 'antd';
import Scrollbars from 'react-custom-scrollbars';
import { middlewareCaching } from '../../../Redux/actions/publicActions/middleware';
import modelContext from '../../../Models/context';
import ObserverTime from '../../ObserverTime';
import TitleModule from '../../TitleModule';
const { Panel } = Collapse;

class SettingsModule extends React.PureComponent {
  state = {
    haveChanges: [],
    showScrollbar: false,
    emailValue: null,
    isLoadingLogs: false,
    telValue: null,
    newPassword: '',
    oldPassword: '',
    mail: false,
    phone: false,
  };

  static propTypes = {
    onErrorRequstAction: PropTypes.func.isRequired,
    path: PropTypes.string.isRequired,
    onSaveComponentState: PropTypes.func.isRequired,
    router: PropTypes.object.isRequired,
  };

  static contextType = modelContext;

  static getDerivedStateFromProps = (props, state) => {
    const { udata: { email = null, phone = null } = {} } = props;

    if (_.isNull(state.emailValue) && _.isNull(state.telValue)) {
      return {
        ...state,
        emailValue: email,
        telValue: phone,
      };
    }
    return state;
  };

  componentDidMount = () => {
    const { isLoadingLogs = false } = this.state;
    const { router, path, settingsLogs = [], udata: { _id: uid = '' } = {}, onCaching } = this.props;

    if (router && router.routeData[path] && router.routeData[path].haveChanges) {
      this.setState({ ...this.state, ...router.routeData[path] });
    }

    if (!Object.keys(settingsLogs).length && !isLoadingLogs && onCaching) {
      onCaching({
        uid,
        actionType: 'get_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });
    }
  };

  componentDidUpdate = (props, state) => {
    const { showScrollbar, emailValue = '', telValue = '', isLoadingLogs = false } = this.state;
    const {
      settingsLogs = [],
      udata: { _id: uid = '' } = {},
      onCaching,
      shouldUpdate = false,
      onSetStatus,
    } = this.props;

    if (onCaching && ((!Object.keys(settingsLogs).length && !isLoadingLogs) || shouldUpdate)) {
      onCaching({
        uid,
        actionType: 'get_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });
      onSetStatus(false);
      this.setState({
        ...this.state,
        isLoadingLogs: true,
      });
    }

    if (this.refWrapper && this.refColumn) {
      const heightWrapper = this.refWrapper.getBoundingClientRect().height;
      const heightColumn = this.refColumn.getBoundingClientRect().height;

      const isShow = heightWrapper > heightColumn;

      if (isShow !== showScrollbar) {
        return this.setState({
          showScrollbar: isShow,
        });
      }
    } else if (!_.isNull(emailValue) || !_.isNull(telValue)) {
      this.setState({ emailValue: null, telValue: null });
    }
  };

  componentWillUnmount = () => {
    const { haveChanges } = this.state;
    const { onSaveComponentState, path, router } = this.props;
    if (haveChanges && !_.isEqual(router.routeData[path], this.state))
      return onSaveComponentState({ ...this.state, path: path });
  };

  hideMail = event => {
    if (event !== this.state.mail) {
      this.setState({
        ...this.state,
        haveChanges: true,
        mail: event,
      });
    }
  };

  hidePhone = event => {
    if (event !== this.state.phone) {
      this.setState({
        ...this.state,
        haveChanges: true,
        phone: event,
      });
    }
  };

  onChangeInput = ({ target = {} }, key) => {
    const { haveChanges = [] } = this.state;
    const filterChanges = [...haveChanges];
    if (!filterChanges.includes(key)) {
      filterChanges.push(key);
    }

    if (target?.dataset?.id === 'email') {
      this.setState({
        ...this.state,
        haveChanges: !filterChanges.length ? [key] : filterChanges,
        emailValue: target.value,
      });
    } else if (target?.dataset?.id === 'tel') {
      this.setState({
        ...this.state,
        haveChanges: !filterChanges.length ? [key] : filterChanges,
        telValue: target.value,
      });
    } else if (target?.dataset?.id === 'newPassword') {
      this.setState({
        ...this.state,
        haveChanges: !filterChanges.length ? [key] : filterChanges,
        newPassword: target.value,
      });
    } else if (target?.dataset?.id === 'oldPassword') {
      this.setState({
        ...this.state,
        haveChanges: !filterChanges.length ? [key] : filterChanges,
        oldPassword: target.value,
      });
    }
  };

  onSaveSettings = async (event, settingsKey) => {
    if (settingsKey.includes('password')) {
      this.onChangePassword(settingsKey);
    } else if (settingsKey === 'common') {
      this.onChangeCommon(settingsKey);
    }
  };

  onChangeCommon = async keyChange => {
    try {
      const { Request = {} } = this.context;
      const { udata: { _id: uid = '' } = {}, onUpdateUdata = null, onCaching = null } = this.props;
      const { emailValue: newEmail = '', telValue: newPhone = '', haveChanges } = this.state;

      if (!newEmail || !/\w+@\w+\.\D+/i.test(newEmail)) {
        message.error('Формат почты не соблюден');
        return;
      }

      if (!uid) {
        message.error('Пользователь не найден');
        return;
      }

      const queryParams = {
        newEmail,
        newPhone,
        uid,
      };
      const rest = new Request();
      const res = await rest.sendRequest('/settings/common', 'POST', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request change password');
      }

      if (onUpdateUdata) {
        onUpdateUdata({ email: newEmail, phone: newPhone });
      }

      this.setState({
        haveChanges: this.state.haveChanges.filter(it => {
          if (it !== keyChange) return true;
          else return false;
        }),
      });

      const isOnlyPhone = haveChanges.includes('commonPhone') && !haveChanges.includes('commonEmail');
      const isOnlyEmail = haveChanges.includes('commonEmail') && !haveChanges.includes('commonPhone');

      const msg = isOnlyPhone
        ? 'Телефон обновлен.'
        : isOnlyEmail
        ? 'Почта обновлена.'
        : 'Почта и телефон обновлены.';

      if (!onCaching) return;

      onCaching({
        uid,
        item: {
          uid,
          date: new Date(),
          message: msg,
          depKey: 'settings',
        },
        actionType: 'save_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });

      this.setState(
        {
          ...this.state,
          haveChanges: [],
        },
        () => {
          message.success('Настройки успешно обновлены.');
        },
      );
    } catch (error) {
      console.error(error);
      message.error('Ошибка смены пароля');
    }
  };

  onChangePassword = async keyChange => {
    try {
      const { Request = {} } = this.context;
      const { udata: { _id: uid = '' } = {}, onCaching } = this.props;
      const { oldPassword = '', newPassword = '' } = this.state;
      if (!oldPassword || !newPassword) {
        message.warning('Формат пароля не верен');
        return;
      }

      if (!uid) {
        message.error('Пользователь не найден');
        return;
      }

      const queryParams = {
        oldPassword,
        newPassword,
        uid,
      };
      const rest = new Request();
      const res = await rest.sendRequest('/settings/password', 'POST', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request change password');
      }

      this.setState({
        haveChanges: this.state.haveChanges.filter(it => {
          if (it !== keyChange) return true;
          else return false;
        }),
      });

      const msg = 'Изменение пароля.';

      if (!onCaching) return;

      onCaching({
        uid,
        item: {
          uid,
          date: new Date(),
          message: msg,
          depKey: 'settings',
        },
        actionType: 'save_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });

      this.setState(
        {
          ...this.state,
          haveChanges: [],
        },
        () => {
          message.success('Пароль изменен.');
        },
      );
    } catch (error) {
      console.error(error);
      message.error('Ошибка смены пароля');
    }
  };

  refWrapper = null;
  refColumn = null;
  refFunc = node => (this.refWrapper = node);
  refColumnFunc = node => (this.refColumn = node);

  render() {
    const { emailValue, telValue, haveChanges, oldPassword, newPassword } = this.state;
    const { settingsLogs = {}, shouldUpdate } = this.props;
    const text = ` A dog is a type of domesticated animal.`;

    const readonlyPassword = haveChanges.includes('password_new') && haveChanges.includes('password_old');
    const readonlyCommon = haveChanges.includes('commonEmail') || haveChanges.includes('commonPhone');

    const settingsBlock = (
      <React.Fragment>
        <div ref={this.refFunc}>
          <Collapse defaultActiveKey={['common']}>
            <Panel onChange={this.onUpdate} header="Смена пароля" key="password">
              <div className="configWrapper flexWrapper">
                <span>Старый пароль:</span>
                <Input
                  data-id="oldPassword"
                  type="password"
                  allowClear
                  value={oldPassword}
                  onChange={e => this.onChangeInput(e, 'password_old')}
                />
              </div>
              <div className="configWrapper flexWrapper">
                <span>Новый пароль:</span>
                <Input
                  data-id="newPassword"
                  type="password"
                  allowClear
                  value={newPassword}
                  onChange={e => this.onChangeInput(e, 'password_new')}
                />
              </div>
              <Button
                onClick={e => this.onSaveSettings(e, 'password')}
                className="submit"
                type="primary"
                disabled={!readonlyPassword}
              >
                Принять изменения
              </Button>
            </Panel>
            <Panel onChange={this.onUpdate} header="Общие настройки" key="common">
              <div className="settingsPanel--center-flex">
                <div className="configWrapper flexWrapper">
                  <span>Сменить почту:</span>
                  <Input
                    data-id="email"
                    allowClear
                    value={emailValue}
                    onChange={e => this.onChangeInput(e, 'commonEmail')}
                  />
                </div>
                <div className="configWrapper flexWrapper">
                  <span>Сменить телефон:</span>
                  <Input
                    data-id="tel"
                    type="tel"
                    allowClear
                    value={telValue}
                    onChange={e => this.onChangeInput(e, 'commonPhone')}
                  />
                </div>
              </div>
              <Button
                onClick={e => this.onSaveSettings(e, 'common')}
                className="submit"
                type="primary"
                disabled={!readonlyCommon}
              >
                Принять изменения
              </Button>
            </Panel>

            <Panel header="Настройки профиля" key="profile">
              <div className="configWrapper">
                <Switch defaultChecked={this.state.mail} onChange={this.hideMail} />
                <span className="configTitle">Скрывать почту</span>
              </div>
              <div className="configWrapper">
                <Switch defaultChecked={this.state.phone} onChange={this.hidePhone} />
                <span className="configTitle">Скрывать телефон</span>
              </div>
              <Button
                onClick={e => this.onSaveSettings(e, 'profile')}
                className="submit"
                type="primary"
                disabled={!haveChanges.includes('profile')}
              >
                Принять изменения
              </Button>
            </Panel>
            <Panel header="Настройки уровней доступа" key="access">
              <Collapse bordered={false}>
                <Panel header="Администратор" key="admin">
                  {text}
                </Panel>
                <Panel header="Начальник отдела" key="headDepartament">
                  {text}
                </Panel>
                <Panel header="Сотрудник" key="solider">
                  {text}
                </Panel>
              </Collapse>
              <Button
                onClick={e => this.onSaveSettings(e, 'access')}
                className="submit"
                type="primary"
                disabled={!haveChanges.includes('profile')}
              >
                Принять изменения
              </Button>
            </Panel>
          </Collapse>
        </div>
      </React.Fragment>
    );

    return (
      <div className="settingsModule">
        <TitleModule classNameTitle="settingsModuleTitle" title="Настройки" />
        <div className="settingsModule__main">
          <div ref={this.refColumnFunc} className="col-6">
            <Scrollbars>{settingsBlock}</Scrollbars>
          </div>
          <div className="col-6">
            <ObserverTime isLoading={!shouldUpdate && !settingsLogs?.length} settingsLogs={settingsLogs} />
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => {
  const {
    publicReducer: { udata = {}, caches = {} } = {},
    router: { shouldUpdate = false, currentActionTab = '' } = {},
  } = state;

  let settingsLogs = [];
  const filterLogs = Object.keys(caches).reduce((logs, key) => {
    if (key.includes('user_settings_log') && !_.isEmpty(caches[key])) {
      logs[key] = { ...caches[key] };
    }
    return logs;
  }, {});

  if (!_.isEmpty(filterLogs)) {
    settingsLogs = Object.keys(filterLogs)
      .map(logKey => {
        return {
          ...filterLogs[logKey],
          date: new Date(filterLogs[logKey].date),
        };
      })
      .sort((a, b) => b.date - a.date);
  }

  return {
    router: { ...state.router },
    settingsLogs,
    udata,
    shouldUpdate: shouldUpdate && currentActionTab.includes('settings'),
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onSaveComponentState: data => dispatch(saveComponentStateAction(data)),
    onUpdateUdata: payload => dispatch(updateUdata(payload)),
    onCaching: props => dispatch(middlewareCaching(props)),
    onSetStatus: props => dispatch(setStatus(props)),
  };
};

export { SettingsModule };
export default connect(mapStateToProps, mapDispatchToProps)(SettingsModule);
