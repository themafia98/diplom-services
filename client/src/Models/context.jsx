import { createContext } from 'react';
import config from '../config.json';
import Request from '../Models/Rest';
import Schema from '../Models/Schema';
import Notification from '../Models/NotificationRouter';
import { clientDB } from './ClientSideDatabase';

export const modelMethods = {
  clientDB,
  rest: new Request(),
  Request,
  config,
  schemaModel: Schema,
  schema: new Schema('no-strict'),
  Notification,
};

const modelsContext = createContext(modelMethods);

export default modelsContext;
