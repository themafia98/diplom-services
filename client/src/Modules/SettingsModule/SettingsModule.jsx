// @ts-nocheck
import React from 'react';
import { settingsModuleType } from './types';
import _ from 'lodash';
import { connect } from 'react-redux';
import {
  settingsLogsSelector,
  settingsStatusSelector,
  settingsArtifactsSelector,
} from '../../Utils/selectors';
import { saveComponentStateAction } from '../../Redux/actions/routerActions';
import { updateUdata, setStatus, onLoadSettings } from '../../Redux/actions/publicActions';
import { message } from 'antd';
import Scrollbars from 'react-custom-scrollbars';
import { middlewareCaching } from '../../Redux/actions/publicActions/middleware';
import modelContext from '../../Models/context';
import ObserverTime from '../../Components/ObserverTime';
import TitleModule from '../../Components/TitleModule';

import PanelPassword from './Panels/PanelPassword';
import PanelCommon from './Panels/PanelCommon';
import PanelAdmin from './Panels/PanelAdmin';
import PanelProfile from './Panels/PanelProfile';

class SettingsModule extends React.PureComponent {
  state = {
    showScrollbar: false,
    emailValue: null,
    isLoadingLogs: false,
    telValue: null,
    newPassword: '',
    oldPassword: '',
    isHideEmail: null,
    isHidePhone: null,
    statusList: [],
  };

  static propTypes = settingsModuleType;
  static contextType = modelContext;
  static defaultProps = {
    udata: {},
    settingsLogs: [],
  };

  refWrapper = null;
  refColumn = null;
  refFunc = (node) => (this.refWrapper = node);
  refColumnFunc = (node) => (this.refColumn = node);

  static getDerivedStateFromProps = (props, state) => {
    const {
      udata: {
        email: emailValue = null,
        phone: telValue = null,
        isHideEmail = null,
        isHidePhone = null,
      } = {},
    } = props;
    const enumState = [state.emailValue, state.telValue, state.isHideEmail, state.isHidePhone];
    if (enumState.every((value) => _.isNull(value))) {
      return {
        ...state,
        emailValue,
        telValue,
        isHideEmail,
        isHidePhone,
      };
    }
    return state;
  };

  componentDidMount = () => {
    const { isLoadingLogs = false } = this.state;
    const { router, path, settingsLogs = [], udata: { _id: uid = '' } = {}, onCaching } = this.props;

    if (router && router.routeData[path] && router.routeData[path].haveChanges) {
      this.setState({ ...this.state, ...router.routeData[path] });
    }

    if (!Object.keys(settingsLogs).length && !isLoadingLogs && onCaching) {
      onCaching({
        uid,
        actionType: 'get_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });
    }
  };

  componentDidUpdate = (props, state) => {
    const { showScrollbar, emailValue = '', telValue = '', isLoadingLogs = false } = this.state;
    const {
      settingsLogs = [],
      udata: { _id: uid = '' } = {},
      onCaching,
      shouldUpdate = false,
      onSetStatus,
    } = this.props;

    if (onCaching && ((!Object.keys(settingsLogs).length && !isLoadingLogs) || shouldUpdate)) {
      onCaching({
        uid,
        actionType: 'get_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });
      onSetStatus(false);
      this.setState({
        ...this.state,
        isLoadingLogs: true,
      });
    }

    if (this.refWrapper && this.refColumn) {
      const heightWrapper = this.refWrapper.getBoundingClientRect().height;
      const heightColumn = this.refColumn.getBoundingClientRect().height;

      const isShow = heightWrapper > heightColumn;

      if (isShow !== showScrollbar) {
        return this.setState({
          showScrollbar: isShow,
        });
      }
    } else if (!_.isNull(emailValue) || !_.isNull(telValue)) {
      this.setState({ emailValue: null, telValue: null });
    }
  };

  onChangeStatusList = async (state, callback = null) => {
    try {
      const { Request } = this.context;
      const { onSaveStatusList } = this.props;

      const queryParams = {
        idSettings: 'statusSettings',
        items: state,
      };

      const rest = new Request();
      const res = await rest.sendRequest('/settings/statusList', 'PUT', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request profile settings');
      }

      const {
        data: {
          response: { metadata: { idSettings: saveSettingsId = '', settings = [] } = {}, metadata = {} } = {},
        } = {},
      } = res;

      if (saveSettingsId !== queryParams.idSettings) {
        throw new Error('Invalid saved settings id');
      }
      onSaveStatusList({ metadata, type: 'update', depKey: saveSettingsId });
      if (callback) callback(settings);
    } catch (error) {
      if (error?.status !== 404) console.error(error);
    }
  };

  onSaveSettings = (settingsKey = '', state = {}, callback = null) => {
    switch (settingsKey) {
      case 'password':
        return this.onChangePassword(state, callback);
      case 'common':
        return this.onChangeCommon(state, callback);
      case 'profile':
        return this.onChangeProfile(state, callback);
      case 'statusSettings':
        return this.onChangeStatusList(state, callback);

      default:
        return;
    }
  };

  onChangeProfile = async (state, callback) => {
    try {
      const { udata: { _id: uid = '' } = {}, onUpdateUdata = null } = this.props;
      const { isHideEmail = false, isHidePhone = false } = state;
      const { config: { settings: { includeChangeProfile = false } = {} } = {} } = this.context;
      const { Request } = this.context;

      if (!includeChangeProfile) return;

      if (!uid) {
        message.error('Пользователь не найден');
        return;
      }

      const queryParams = {
        isHideEmail,
        isHidePhone,
        uid,
      };
      const rest = new Request();
      const res = await rest.sendRequest('/settings/profile', 'POST', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request profile settings');
      }

      if (onUpdateUdata) {
        onUpdateUdata({ isHideEmail, isHidePhone });
      }

      if (callback) callback();
    } catch (error) {
      if (error?.status !== 404) console.error(error);
    }
  };

  onChangeCommon = async (state, callback) => {
    try {
      const { Request = {} } = this.context;
      const { udata: { _id: uid = '' } = {}, onUpdateUdata = null, onCaching = null } = this.props;
      const { emailValue: newEmail = '', telValue: newPhone = '', haveChanges = [] } = state;
      const { config: { settings: { includeChangeEmail = false } = {} } = {} } = this.context;
      if (includeChangeEmail && (!newEmail || !/\w+@\w+\.\D+/i.test(newEmail))) {
        message.error('Формат почты не соблюден');
        return;
      }

      if (!uid) {
        message.error('Пользователь не найден');
        return;
      }

      const queryParams = includeChangeEmail
        ? {
            newEmail,
            newPhone,
            uid,
          }
        : { newPhone, uid };
      const rest = new Request();
      const res = await rest.sendRequest('/settings/common', 'POST', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request change password');
      }

      if (onUpdateUdata) {
        onUpdateUdata(includeChangeEmail ? { email: newEmail, phone: newPhone } : { phone: newPhone });
      }

      if (callback) callback();

      const isOnlyPhone = haveChanges.includes('commonPhone') && !haveChanges.includes('commonEmail');
      const isOnlyEmail = haveChanges.includes('commonEmail') && !haveChanges.includes('commonPhone');

      const msg = isOnlyPhone
        ? 'Телефон обновлен.'
        : isOnlyEmail
        ? 'Почта обновлена.'
        : 'Почта и телефон обновлены.';

      if (!onCaching) return;

      onCaching({
        uid,
        item: {
          uid,
          date: new Date(),
          message: msg,
          depKey: 'settings',
        },
        actionType: 'save_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });

      message.success('Настройки успешно обновлены.');
    } catch (error) {
      if (error?.response?.status !== 404) console.error(error);
      message.error('Ошибка смены пароля');
    }
  };

  onChangePassword = async (state, callback) => {
    try {
      const { Request = {} } = this.context;
      const { udata: { _id: uid = '' } = {}, onCaching } = this.props;
      const { oldPassword = '', newPassword = '' } = state;
      if (!oldPassword || !newPassword) {
        message.warning('Формат пароля не верен');
        return;
      }

      if (!uid) {
        message.error('Пользователь не найден');
        return;
      }

      const queryParams = {
        oldPassword,
        newPassword,
        uid,
      };
      const rest = new Request();
      const res = await rest.sendRequest('/settings/password', 'POST', { queryParams }, true);

      if (!res || res.status !== 200) {
        throw new Error('Bad request change password');
      }

      if (callback) callback();

      const msg = 'Изменение пароля.';

      if (!onCaching) return;

      onCaching({
        uid,
        item: {
          uid,
          date: new Date(),
          message: msg,
          depKey: 'settings',
        },
        actionType: 'save_user_settings_log',
        depStore: 'settings',
        type: 'logger',
      });

      message.success('Пароль изменен.');
    } catch (error) {
      if (error?.response?.status !== 404) console.error(error);
      message.error('Ошибка смены пароля');
    }
  };

  render() {
    const {
      emailValue,
      telValue,
      oldPassword,
      newPassword,
      isHideEmail,
      isHidePhone,
      isLoadingLogs = false,
    } = this.state;
    const { settingsLogs = null, udata: { departament = '', rules = '' } = {}, settings = [] } = this.props;
    const { config: { settings: { includeRulesSettings = false } = {} } = {} } = this.context;

    const isLoading = !isLoadingLogs && (!settingsLogs || (settingsLogs && !settingsLogs?.length));
    const isAdmin = departament === 'Admin' && rules === 'full';
    return (
      <div className="settingsModule">
        <TitleModule classNameTitle="settingsModuleTitle" title="Настройки" />
        <div className="settingsModule__main">
          <div ref={this.refColumnFunc} className="col-6">
            <Scrollbars>
              <div ref={this.refFunc}>
                <PanelPassword
                  oldPassword={oldPassword}
                  newPassword={newPassword}
                  onSaveSettings={this.onSaveSettings}
                />
                <PanelCommon
                  onSaveSettings={this.onSaveSettings}
                  emailValue={emailValue}
                  telValue={telValue}
                />
                <PanelProfile
                  onSaveSettings={this.onSaveSettings}
                  isHideEmail={isHideEmail}
                  isHidePhone={isHidePhone}
                />
                {isAdmin && includeRulesSettings ? (
                  <PanelAdmin statusList={settings} onSaveSettings={this.onSaveSettings} />
                ) : null}
              </div>
            </Scrollbars>
          </div>
          <div className="col-6">
            <ObserverTime isLoading={isLoading} settingsLogs={settingsLogs} />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const {
    publicReducer: { udata = {} } = {},
    router: { shouldUpdate = false, currentActionTab = '' } = {},
    router = {},
  } = state;

  return {
    router,
    udata,
    settings: settingsStatusSelector(state, props),
    artifacts: settingsArtifactsSelector(state, props),
    settingsLogs: settingsLogsSelector(state, props),
    shouldUpdate: shouldUpdate && currentActionTab.includes('settings'),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onSaveComponentState: (data) => dispatch(saveComponentStateAction(data)),
    onUpdateUdata: (payload) => dispatch(updateUdata(payload)),
    onCaching: (props) => dispatch(middlewareCaching(props)),
    onSetStatus: (props) => dispatch(setStatus(props)),
    onSaveStatusList: (payload) => dispatch(onLoadSettings(payload)),
  };
};

export { SettingsModule };
export default connect(mapStateToProps, mapDispatchToProps)(SettingsModule);
