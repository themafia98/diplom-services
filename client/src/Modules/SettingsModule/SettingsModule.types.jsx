import PropTypes from 'prop-types';
import { udataType } from 'App.types';
const { func, bool, object, string, array } = PropTypes;

export const settingsModuleType = {
  visible: bool,
  rest: object.isRequired,
  path: string.isRequired,
  settingsLogs: array.isRequired,
  udata: udataType.isRequired,
  shouldUpdate: bool.isRequired,
  onSaveComponentState: func.isRequired,
  onUpdateUdata: func.isRequired,
  onCaching: func.isRequired,
  onSetStatus: func.isRequired,
};
