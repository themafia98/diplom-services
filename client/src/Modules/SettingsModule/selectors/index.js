// @ts-nocheck
import { createSelector } from 'reselect';
import _ from 'lodash';
import { cache } from './selectors';

const settingsLogsSelector = createSelector(cache, (caches) => {
  const cahceLogs = Object.keys(caches).reduce((logs, key) => {
    if (key.includes('user_settings_log') && !_.isEmpty(caches[key])) {
      logs[key] = { ...caches[key] };
    }
    return logs;
  }, {});

  if (_.isEmpty(cahceLogs)) return [];

  const settingsLogs = Object.keys(cahceLogs)
    .map((logKey) => {
      return {
        ...cahceLogs[logKey],
        date: new Date(cahceLogs[logKey].date),
      };
    })
    .sort((a, b) => b.date - a.date);

  return settingsLogs;
});

export { settingsLogsSelector };
