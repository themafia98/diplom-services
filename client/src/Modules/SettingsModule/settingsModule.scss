.settingsModule {
  .settingsModuleTitle {
    font-size: 30px;
    font-weight: bold;
    margin: 0;
    user-select: none;
  }
  .settingsSelect-status {
    width: 100%;
  }

  .settingsModule__language {
    margin-top: 5px;
    .ant-select {
      width: 100%;
    }
  }

  .settingsPanel--center-flex {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }
  .flexWrapper {
    display: flex;
    width: 70%;
    justify-content: center;
    span:first-child {
      font-weight: bold;
      flex: 0 0 25%;
    }
    input {
      flex: 0 0 75%;
      text-align: center;
    }
  }
  .submit {
    margin: 10px 3px;
  }

  .settingsModule__main {
    display: flex;
    height: calc(100vh - 180px);
    .configWrapper {
      display: flex;
      align-items: center;
      margin-bottom: 10px;
      .configTitle {
        padding-left: 10px;
      }
    }
    .ant-collapse-content-box {
      .ant-collapse-item {
        border: none;
      }
    }
  }
}
