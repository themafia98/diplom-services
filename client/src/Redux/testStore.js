import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

/** stores for  jest */

export const initialState = {
  tabReducer: {
    parentSize: null,
    childrenSize: null,
    flag: false,
  },
  publicReducer: {
    status: 'online',
    prewStatus: 'online',
    requestError: null,
    caches: {},
  },
  router: {
    currentActionTab: 'mainModule',
    actionTabs: ['MainModule'],
    routeDataActive: {},
    routeData: {},
  },
};

const middlewares = [thunk.withExtraArgument({})]; /** test middlewares */
const mockStore = configureStore(middlewares); /** test config store */

const store = mockStore(initialState); /** init test store */

export default store;
