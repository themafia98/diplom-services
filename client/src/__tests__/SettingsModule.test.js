import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { initialState } from '../Redux/testStore';
import { SettingsModule } from '../Modules/SettingsModule/SettingsModule';
import context from '../Models/context';
describe('<SettingsModule /> template', () => {
  test('Should be render', () => {
    const props = {
      onErrorRequestAction: () => {},
      path: 'settingsModule',
      statusApp: 'online',
      visible: true,
      loaderMethods: {},
      rest: {},
      shouldUpdate: false,
      onUpdateUdata: () => {},
      onCaching: () => {},
      onSetStatus: () => {},
      emailValue: '',
      onSaveCache: () => {},
      telValue: '',
      isHidePhone: false,
      isHideEmail: false,
      onSaveComponentState: () => {},
      router: { ...initialState.router },
    };

    const SettingsModuleWrapper = shallow(<SettingsModule {...props} />, { context });
    expect(toJson(SettingsModuleWrapper)).toMatchSnapshot();

    /** TODO: need upgrade test for new settings component */
    // expect(
    //   SettingsModuleWrapper.findWhere(node => node.at(0) && node.hasClass('submit')).prop('disabled'),
    // ).toEqual(true);
    // expect(
    //   SettingsModuleWrapper.findWhere(node => node.at(0) && node.hasClass('submit')).simulate('click', []),
    // ).toBeTruthy();

    const arrayPanel = SettingsModuleWrapper.find('Collapse');

    arrayPanel.forEach((node, i) => {
      if (i === 0) expect(node.prop('bordered')).toEqual(true);
      else expect(node.prop('bordered')).toEqual(false);

      expect(node.simulate('click', []));
    });

    SettingsModuleWrapper.setState({ hasChange: true });
    expect(SettingsModuleWrapper.find('.submit')).toEqual({});

    expect(toJson(SettingsModuleWrapper)).toMatchSnapshot();
  });
});
