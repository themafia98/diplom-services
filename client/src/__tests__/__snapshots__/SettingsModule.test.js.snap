// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsModule /> template Should be render 1`] = `
<div
  className="settingsModule"
>
  <TitleModule
    additional=""
    className=""
    classNameTitle="settingsModuleTitle"
    title="Настройки"
  />
  <div
    className="settingsModule__main"
  >
    <div
      className="col-6"
    >
      <Scrollbars
        autoHeight={false}
        autoHeightMax={200}
        autoHeightMin={0}
        autoHide={false}
        autoHideDuration={200}
        autoHideTimeout={1000}
        hideTracksWhenNotNeeded={false}
        renderThumbHorizontal={[Function]}
        renderThumbVertical={[Function]}
        renderTrackHorizontal={[Function]}
        renderTrackVertical={[Function]}
        renderView={[Function]}
        tagName="div"
        thumbMinSize={30}
        universal={false}
      >
        <div>
          <PanelPassword
            newPassword=""
            oldPassword=""
            onSaveSettings={[Function]}
          />
          <PanelCommon
            emailValue={null}
            onSaveSettings={[Function]}
            telValue={null}
          />
          <PanelProfile
            isHideEmail={null}
            isHidePhone={null}
            onSaveSettings={[Function]}
          />
        </div>
      </Scrollbars>
    </div>
    <div
      className="col-6"
    >
      <ObserverTime
        isLoading={true}
        settingsLogs={Array []}
        title=""
      />
    </div>
  </div>
</div>
`;

exports[`<SettingsModule /> template Should be render 2`] = `
<div
  className="settingsModule"
>
  <TitleModule
    additional=""
    className=""
    classNameTitle="settingsModuleTitle"
    title="Настройки"
  />
  <div
    className="settingsModule__main"
  >
    <div
      className="col-6"
    >
      <Scrollbars
        autoHeight={false}
        autoHeightMax={200}
        autoHeightMin={0}
        autoHide={false}
        autoHideDuration={200}
        autoHideTimeout={1000}
        hideTracksWhenNotNeeded={false}
        renderThumbHorizontal={[Function]}
        renderThumbVertical={[Function]}
        renderTrackHorizontal={[Function]}
        renderTrackVertical={[Function]}
        renderView={[Function]}
        tagName="div"
        thumbMinSize={30}
        universal={false}
      >
        <div>
          <PanelPassword
            newPassword=""
            oldPassword=""
            onSaveSettings={[Function]}
          />
          <PanelCommon
            emailValue={null}
            onSaveSettings={[Function]}
            telValue={null}
          />
          <PanelProfile
            isHideEmail={null}
            isHidePhone={null}
            onSaveSettings={[Function]}
          />
        </div>
      </Scrollbars>
    </div>
    <div
      className="col-6"
    >
      <ObserverTime
        isLoading={false}
        settingsLogs={Array []}
        title=""
      />
    </div>
  </div>
</div>
`;
